sdk/iot/core/doc/mqtt_state_machine.md

```diff
@@ -136,3 +138,120 @@ _Example:_
+
+#### Error policy
+
+Our SDK will not handle protocol-level (WebSocket, MQTT, TLS or TCP) errors. The application-developer is expected to classify and handle errors the following way:
+
+- Authentication errors should not be retried.
+- Communication-related errors other than ones security-related should be considered retriable.
+
+Both IoT Hub and Provisioning services will use `MQTT CONNACK` as described in Section 3.2.2.3 of the [MQTT v3 specification](http://docs.oasis-open.org/mqtt/mqtt/v3.1.1/os/mqtt-v3.1.1-os.html#_Table_3.1_-).
+
+##### IoT Hub Service Errors
+
+APIs using `az_iot_hub_client_status` report service-side errors to the client through the IoT Hub protocols. (At the time of writing, only Twin responses may return errors.)
+
+The following APIs may be used to determine if the status indicates error and if the error should be retried:
+
+```C
+az_iot_hub_client_status status = response.status;
+if (az_iot_hub_client_is_success_status(status))
+{
+    // success case 
+}
+else
+{
+    if (az_iot_hub_client_is_retriable_status(status))
+    {
+        // retry
+    }
+    else
+    {
+        // fail
+    }
+}
+```
+
+##### Provisioning Service Errors
+
+Each Provisioning operation will return a status encoded as `az_iot_provisioning_client_status`.
+After the register operation is complete the `az_iot_provisioning_client_registration_state registration_information` field will contain the `az_iot_provisioning_client_status error_code` that represents the overall status of the registration operation.  
+
+The following APIs may be used to determine if the status indicates error and if the error should be retried:
+
+```C
+az_iot_provisioning_client_status status = result.registration_information.error_code;
+if (az_iot_provisioning_client_is_success_status(status))
+{
+    // success case 
+}
+else
+{
+    if (az_iot_provisioning_client_is_retriable_status(status))
+    {
+        // retry
+    }
+    else
+    {
+        // fail
+    }
+}
+```
+
+#### Retry timing
+
+Network timeouts and the MQTT keep-alive interval should be configured considering tradeoffs between how fast network issues are detected vs traffic overheads. [This document](https://docs.microsoft.com/en-us/azure/iot-hub/iot-hub-mqtt-support#default-keep-alive-timeout) describes the recommended keep-alive timeouts as well as the minimum idle timeout supported by Azure IoT services.
+
+For connectivity issues at all layers (TCP, TLS, MQTT) as well as cases where there is no `retry-after` sent by the service, we suggest using an exponential back-off with random jitter function. `_az_retry_calc_delay` is available in Azure SDK Core:
+
+```C
+// random_ms is a random value between 0 and max_jitter_msec
+// The previous operation took operation_msec
+
+int32_t total_delay_msec =  random_msec + _az_retry_calc_delay(attempt, min_retry_delay_msec, max_retry_delay_msec);

```

Comment:
@JeffreyRichter @antkmsft Please review this portion: 
1. `_az_retry_calc_delay` doesn't include a random jitter. This is fine as the application may have platform-specific ways of implementing that. Should it though be included in the calculation (to have a strong requirement for a random int)?
2. The API above is "internal" and I'm recommending it is used by app-developers. We do expect it to be mostly used by convenience layers but the Embedded devices IoT SDK should be usable as-is.

Reply:
If you expect customers code to call this (under ANY circumstance [like not using one of your convenience layers), then this function should be public, documents, tested, fully supported, and NOT start with an underscore. 

I don't think that jitter is important. We have a strong goal of keeping the PAL functions to a minimum. Everything the customer has to do to bring up the SDK on a new device is a barrier to entry so the fewer functions the better. How bout we ignore jitter for now and then address it IF it becomes a problem later,

Reply:
I do expect customer code to call this since there is nothing else similar to it. Should I:

1. Propose a change in SDKs to make this public
1. Make our own az_iot specific public function (in which case I'll likely add a mandatory jitter "random" arg).

Reply:
Yes, make this public in your client library. You can add the jitter param.